test/common/testlib.py

```diff
@@ -654,6 +661,63 @@ def snapshot(self, title, label=None):
             attach(filename)
             print("Wrote HTML dump to " + filename)
 
+    def assert_pixels(self, selector, key):
+        """Compare the given element with its reference"""
+
+        if Image and self.pixels_label and self.cdp and self.cdp.valid:
+            rect = self.call_js_func('ph_element_clip', selector)
+            base = self.pixels_label + "-" + key
+            filename = base + "-pixels.png"
+            ref_filename = os.path.join(TEST_DIR, "reference", filename)
+            ret = self.cdp.invoke("Page.captureScreenshot", clip=rect, no_trace=True)
+            if "data" in ret:
+                png_now = base64.standard_b64decode(ret["data"])
+                png_ref = os.path.exists(ref_filename) and open(ref_filename, "rb").read()
+                if not png_ref:
+                    with open(filename, 'wb') as f:
+                        f.write(png_now)
+                    attach(filename)
+                    print("New pixel test reference " + filename)
+                    self.failed_pixel_tests += 1
+                else:
+                    img_now = Image.open(io.BytesIO(png_now)).convert("RGBA")
+                    img_ref = Image.open(io.BytesIO(png_ref)).convert("RGBA")
+                    img_delta = img_ref.copy()
+
+                    def img_eq(ref, now, delta):
+                        # This is slow but exactly what we want.
+                        # ImageMath might be able to speed this up.
+                        if ref.size != now.size:
+                            return False
+                        data_ref = ref.load()
+                        data_now = now.load()
+                        data_delta = delta.load()
+                        result = True
+                        width, height = ref.size
+                        for y in range(height):
+                            for x in range(width):
+                                if data_ref[x, y][3] == 255 and data_ref[x, y] != data_now[x, y]:
+                                    data_delta[x, y] = (255, 0, 0, 255)
+                                    result = False
```

Comment:
Ugh.. hopefully our screenshots are small enough, this looks pretty horrible :-) (We use PIL anyway, doesn't that have such functionality? Or maybe imagemagick/graphicsmagick?)

Reply:
PIL can do some math, but I haven't taken the time yet to figure it all out.  This naive code needs a few seconds per image, which is very slow for what it does, but still acceptable in the big picture, I'd say.